### DISCUSSION TITLE
ca certificates with node fetch?

### DISCUSSION THREAD

**Discussion by everett1992** [nodejs/undici#1364] [Q&A]
I cannot find how to provide ca certificates with `fetch`. Google has no results for ca, certificates, or tls on site:undici.nodejs.org.

Is this missing documentation, or a missing feature? At work different services use different CA's and we shouldn't add all certificates to NODE_EXTRA_CERTS, so we must pass certs with requests.

**Comment by mcollina** [comment]
fetch() does not provide this as part of the spec, therefore it was not added. Would you like to work on this feature? Could you check if it's doable using any other API in undici?

**Reply by everett1992** [reply]
This is something I'm willing to work on. I'm interested in undici `fetch` because it is the global fetch implementation in nodejs so I'm only interested in contributing to the api's exposed there. `ca`  can be provided to a  Client's [connect](https://undici.nodejs.org/#/docs/api/Client?id=parameter-connectoptions) param, 


Is there prior art or discussion around non-spec client configuration for fetch?

The options I can think of are:
1. to include unspecified options on init or Request.
  ```
  fetch('http://example.com', { ca })
  ```
  [node-fetch](https://www.npmjs.com/package/node-fetch#user-content-options) uses this option, accepting an `agent` in fetch options.

2. a fetch constructor that returns a configured spec compliant `fetch` function
  ```
  const client = new Client({ connect: { ca }})
  const ourFetch = client.fetch
  ```

**Reply by mcollina** [reply]
I think the best way is to have a custom, undici-specific dispatcher option (dispatcher is an interface). Then you can instantiate an `Agent`, `Pool`, or `Client` as needed.

**Reply by everett1992** [reply]
How is dispatcher exposed in the fetch interface?

**Reply by mcollina** [reply]
It's currently not exposed.
